pkg/provider/prefix/provider.go

```diff
@@ -0,0 +1,428 @@
+		if job.Operations != worker.OperationReconcileNotRequired {
+			p.SubmitAsyncJob(job)
+		}
+	}
+}
+
+func (p *ipv4PrefixProvider) GetPool(nodeName string) (pool.Pool, bool) {
+	providerAndPool, exists := p.getInstanceProviderAndPool(nodeName)
+	if !exists {
+		return nil, false
+	}
+	return providerAndPool.resourcePool, true
+}
+
+func (p *ipv4PrefixProvider) IsInstanceSupported(instance ec2.EC2Instance) bool {
+	//TODO add check for Nitro instances
+	if instance.Os() == config.OSWindows {
+		return true
+	}
+
+	return false
+}
+
+func (p *ipv4PrefixProvider) Introspect() interface{} {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	response := make(map[string]pool.IntrospectResponse)
+	for nodeName, resource := range p.instanceProviderAndPool {
+		response[nodeName] = resource.resourcePool.Introspect()
+	}
+	return response
+}
+
+func (p *ipv4PrefixProvider) IntrospectNode(node string) interface{} {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	resource, found := p.instanceProviderAndPool[node]
+	if !found {
+		return struct{}{}
+	}
+	return resource.resourcePool.Introspect()
+}
+
+// putInstanceProviderAndPool stores the node's instance provider and pool to the cache
+func (p *ipv4PrefixProvider) putInstanceProviderAndPool(nodeName string, resourcePool pool.Pool, manager eni.ENIManager, capacity int,
+	isPrevPDEnabled bool) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	resource := &ResourceProviderAndPool{
+		eniManager:      manager,
+		resourcePool:    resourcePool,
+		capacity:        capacity,
+		isPrevPDEnabled: isPrevPDEnabled,
+	}
+
+	p.instanceProviderAndPool[nodeName] = resource
+}
+
+// getInstanceProviderAndPool returns the node's instance provider and pool from the cache
+func (p *ipv4PrefixProvider) getInstanceProviderAndPool(nodeName string) (*ResourceProviderAndPool, bool) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	resource, found := p.instanceProviderAndPool[nodeName]
+	return resource, found
+}
+
+// deleteInstanceProviderAndPool deletes the node's instance provider and pool from the cache
+func (p *ipv4PrefixProvider) deleteInstanceProviderAndPool(nodeName string) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	delete(p.instanceProviderAndPool, nodeName)
+}
+
+// getCapacity returns the capacity for IPv4 addresses deconstructed from IPv4 prefixes based on the instance type and the instance os;
+func getCapacity(instanceType string, instanceOs string) int {
+	// Assign only 1st ENIs non-primary IP
+	limits, found := vpc.Limits[instanceType]
+	if !found {
+		return 0
+	}
+	var capacity int
+	if instanceOs == config.OSWindows {
+		capacity = limits.IPv4PerInterface - 1
+	} else {
+		capacity = (limits.IPv4PerInterface - 1) * limits.Interface
+	}
+
+	return capacity
+}
+
+// Retrieve dynamic configuration for prefix delegation from config map, else use default warm pool config
+func (p *ipv4PrefixProvider) getPDWarmPoolConfig() *config.WarmPoolConfig {
+	resourceConfig := make(map[string]config.ResourceConfig)
```

Comment:
nit: do we need to create a map since the variable will take whatever map returned from the functions?

Reply:
I initialized the map here because the following `if-else` block requires `resourceConfig` to be declared first. The alternative is `var resourceConfig map[string]config.ResourceConfig` but that would leave the map uninitialized and have a value of `nil`. But I guess that should be harmless since it will be replace by the results returned from the functions anyway. I can change it to `var resourceConfig map[string]config.ResourceConfig` if that works.

Reply:
yeah, as long as the function returns a map no matter if it succeeds. It should be safe here. If the function returns nil, it is also not safe here.

Reply:
That makes sense. I double checked both functions won't return nil, so it should be safe to declare `var resourceConfig map[string]config.ResourceConfig`. Updated.